src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.cpp

```diff
@@ -215,6 +215,26 @@ void MetadataStorageFromPlainObjectStorageTransaction::removeDirectory(const std
     }
 }
 
+void MetadataStorageFromPlainObjectStorageTransaction::createHardLink(const std::string & path_from, const std::string & path_to)
+{
+    if (metadata_storage.object_storage->isWriteOnce())
```

Comment:
Is it possible to support s3_plain with CopyObject? (not here though)

Reply:
There is a WIP https://github.com/ClickHouse/clickhouse-private/pull/21824 that implements renameExchange